fog/sample-paykit/src/client.rs

```diff
@@ -362,6 +360,232 @@ impl Client {
+        let fog_uris = (&[&self.account_key.change_subaddress()])
+            .iter()
+            .filter_map(|addr| addr.fog_report_url())
+            .map(FogUri::from_str)
+            .collect::<core::result::Result<Vec<_>, _>>()?;
+        let fog_responses = self
+            .fog_report_conn
+            .fetch_fog_reports(fog_uris.into_iter())?;
+        let fog_resolver = FogResolver::new(fog_responses, &self.fog_verifier)?;
+
+        // Check amount found, calculate change
+        let input_amount = input.amount;
+        if offered.token_id != input_amount.token_id || offered.value > input_amount.value {
+            return Err(Error::InsufficientFunds);
+        }
+
+        let change = Amount::new(input_amount.value - offered.value, offered.token_id);
+
+        let (ring, membership_proofs): (Vec<TxOut>, Vec<TxOutMembershipProof>) =
+            ring.into_iter().unzip();
+        let global_indices: Vec<u64> = membership_proofs.iter().map(|proof| proof.index).collect();
+        let input_credentials = input_credentials_helper(
+            input,
+            input_proof,
+            ring,
+            membership_proofs,
+            &self.account_key,
+        )?;
+
+        // TODO: Use the RTHMemoBuilder?
+        let mut sci_builder = SignedContingentInputBuilder::new(
+            block_version,
+            input_credentials,
+            global_indices,
+            fog_resolver,
+            EmptyMemoBuilder::default(),
+        )?;
+
+        sci_builder
+            .add_required_output(requested, &self.account_key.default_subaddress(), rng)
+            .map_err(Error::AddOutput)?;
+
+        let change_destination = ReservedDestination::from(&self.account_key);
+
+        sci_builder
+            .add_required_change_output(change, &change_destination, rng)
+            .map_err(|err| {
+                log::error!(self.logger, "Could not add change due to {:?}", err);
+                Error::AddOutput(err)
+            })?;
+
+        sci_builder.set_tombstone_block(tombstone_block);
+
+        Ok(sci_builder.build(rng)?)
+    }
+
+    /// Builds a transaction that fulfills a swap request, sending all excess
+    /// funds to ourselves and paying fee.
+    ///
+    /// # Arguments
+    /// * `sci` - The swap request we are fulfilling fee.
+    /// * `fee` - The transaction fee to use
+    /// * `rng` - Randomness.
+    pub fn build_swap_transaction<T: RngCore + CryptoRng>(
+        &mut self,
+        mut sci: SignedContingentInput,
+        fee: Amount,
+        rng: &mut T,
+    ) -> Result<Tx> {
+        mc_common::trace_time!(self.logger, "MobileCoinClient.build_swap_transaction");
+
+        // Validate the sci
+        sci.validate()?;
+
+        // Check if its key image alreay landed
+        let res = self.fog_key_image.check_key_images(&[sci.key_image()])?;
+        if res.results[0].key_image_result_code == KeyImageResultCode::Spent as u32 {
+            return Err(Error::SciExpired);
```

Comment:
I'm not sure how useful this is, due to the racy nature of SCIs. It adds some latency to the building, but probably nice for debugging. Nothing actionable, just thinking out loud.

Reply:
yeah i guess this would make more sense in a non-fog client that has the ledger locally

Reply:
noted in code comment